### ISSUE TITLE
ContentType header parser is too strict.

### ISSUE THREAD

**Issue by hochgi** [zio/zio-http#2416]
**Describe the bug**
It seems like content type header parser is too strict.
Opening this issue due to [request on discord](https://discordapp.com/channels/629491597070827530/819703129267372113/1146425677722165300) from @vigoo 

It seems like valid content types like:
https://github.com/prometheus/client_java/blob/a76a65ddb62b50eb2c33a46e98cde4f4479a54a6/simpleclient_common/src/main/java/io/prometheus/client/exporter/common/TextFormat.java#L18

are not accepted by the parser:
<img width="682" alt="Screen Shot 2023-08-30 at 18 03 31" src="https://github.com/zio/zio-http/assets/881075/e1e476fa-51cf-47f2-a381-049eb58963f2">

**Expected behavior**
Parser should adhere to RFC. in fact it's probably not a bad idea to fully implement the [grammar from RFC](https://www.w3.org/Protocols/rfc1341/4_Content-Type.html):
```
Content-Type := type "/" subtype *[";" parameter] 

type :=          "application"     / "audio" 
          / "image"           / "message" 
          / "multipart"  / "text" 
          / "video"           / x-token 

x-token := <The two characters "X-" followed, with no 
           intervening white space, by any token> 

subtype := token 

parameter := attribute "=" value 

attribute := token 

value := token / quoted-string 

token := 1*<any CHAR except SPACE, CTLs, or tspecials> 

tspecials :=  "(" / ")" / "<" / ">" / "@"  ; Must be in 
           /  "," / ";" / ":" / "\" / <">  ; quoted-string, 
           /  "/" / "[" / "]" / "?" / "."  ; to use within 
           /  "="                        ; parameter values
```
using e.g. zio-parser.

Moreover, it seems not all media types are supported either. The hard-coded list defined at `MediaTypes.allMediaTypes` is not extensible, and types like:
https://github.com/prometheus/client_java/blob/a76a65ddb62b50eb2c33a46e98cde4f4479a54a6/simpleclient_common/src/main/java/io/prometheus/client/exporter/common/TextFormat.java#L25
fail to even parse the media type:
<img width="682" alt="Screen Shot 2023-08-30 at 18 42 06" src="https://github.com/zio/zio-http/assets/881075/ffa92621-109d-484c-8256-cb31cc28af1a">

**explicit expectation from RFC**:
> After the type and subtype names, the remainder of the header field is simply a set of parameters, specified in an attribute/value notation. The set of meaningful parameters differs for the different types. 

Yet the parser only allows `charset` & `boundary` parameters (example in link fails due to `version` parameter).

**Additional context**
As a workaround, it's possible to just use `Header.Custom` instead.

**Comment by FabioPinheiro** [comment]
I just found this problem also.
After updating to version 3.0.0-RC2 my endpoint conditions no longer work.
Because `request.header(Header.ContentType)` always returns None.

**Comment by TomTriple** [comment]
If I see that correctly in the BNF a version parameter like e.g. `0.0.4` is in fact not valid (but possibly accepted by many implementations?) as that contains char `.` which explicitly is disallowed (tspecials) to use in `token`. Using a `.` inside a quoted parameter value like `version="0.0.4"` would be valid.

**Comment by hochgi** [comment]
> If I see that correctly in the BNF a version parameter like e.g. `0.0.4` is in fact not valid (but possibly accepted by many implementations?) as that contains char `.` which explicitly is disallowed (tspecials) to use in `token`. Using a `.` inside a quoted parameter value like `version="0.0.4"` would be valid.

@TomTriple I think you're right!
Its still an issue, as zio-http parser will not accept the valid quoted string either,
but probably you can open an issue for @prometheus to fix their content type header.

**Comment by TomTriple** [comment]
> > If I see that correctly in the BNF a version parameter like e.g. `0.0.4` is in fact not valid (but possibly accepted by many implementations?) as that contains char `.` which explicitly is disallowed (tspecials) to use in `token`. Using a `.` inside a quoted parameter value like `version="0.0.4"` would be valid.
> 
> @TomTriple I think you're right! Its still an issue, as zio-http parser will not accept the valid quoted string either, but probably you can open an issue for @prometheus to fix their content type header.

Yes, it's still an issue. I'll see if I can get something working here.

**Comment by TomTriple** [comment]
@hochgi Parsing `MediaType.forContentType("application/openmetrics-text")` works in master!
